src/vantage6_strongaya_general/miscellaneous.py

```diff
@@ -0,0 +1,665 @@
+    This can, for example, be used to address issues as one-to-many relationships skewing missing data computation.
+    """
+
+    def __init__(self, pandas_obj: pd.DataFrame):
+        """
+        Initialise the PredeterminedInfoAccessor.
+
+        Args:
+            pandas_obj (pd.DataFrame): The DataFrame this accessor is attached to
+        """
+        self._obj = pandas_obj
+        self._initialized = False
+
+    def _check_initialized(self) -> None:
+        """
+        Initialize the storage if not already done.
+
+        Ensures the DataFrame has the necessary attribute storage structure.
+        """
+        if not self._initialized:
+            if "stats" not in self._obj.attrs:
+                self._obj.attrs["stats"] = {}
+            self._initialized = True
+
+    def add_stat(
+        self,
+        stat_name: str,
+        calculation_func: Optional[Callable[..., Any]] = None,
+        value: Any = None,
+        per_column: bool = False,
+        store_output_index: Optional[int] = None,
+        update_with_output_index: Optional[int] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Add a custom statistic to the DataFrame.
+
+        Args:
+            stat_name (str): Name of the statistic to store
+            calculation_func (Optional[Callable[..., Any]]): Function to calculate the statistic
+            value (Any, optional): Direct value to store if not using calculation_func
+            per_column (bool): If True, calculate/store the stat for each column separately
+            store_output_index (Optional[int]): If function returns tuple, index of value to store
+            update_with_output_index (Optional[int]): If function returns tuple, index to update DataFrame
+            **kwargs: Additional arguments to pass to calculation_func
+
+        Raises:
+            ValueError: If the value is not JSON serialisable
+        """
+        self._check_initialized()
+        safe_log(
+            "info",
+            f"Adding predetermined information '{stat_name}' to DataFrame attributes",
+        )
+
+        if value is None and calculation_func is not None:
+            if per_column:
+                value = {}
+                updated_columns = {}
+                for col in self._obj.columns:
+
+                    def calculate_for_column(**kwargs: Any) -> Any:
+                        result = calculation_func(self._obj[col], **kwargs)
+                        if isinstance(result, tuple):
+                            stat_value = (
+                                result[store_output_index]
+                                if store_output_index is not None
+                                else result[0]
+                            )
+                            if update_with_output_index is not None:
+                                updated_columns[col] = result[update_with_output_index]
+                            return stat_value
+                        return result
```

Comment:
Please don't define the subfunc within a for loop (I guess Python will redefine it every time?!). I think it's good practice to only use subfunctions when there is a clear goal, here they could also be private class functions